### PR TITLE
Fixed unexpected np.nan value with reindex on pd.series with pd.Inter…

### LINKED ISSUE #51826: BUG:  unexpected value returned when `.reindex`ing on a `pd.Series` with a `pd.IntervalIndex` when using `np.nan` as an index value
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
>>> import pandas as pd
>>> import numpy as np

>>> # expected behavior, currently working
>>> base = 100
>>> d = pd.Series(range(base), index=pd.IntervalIndex.from_arrays(range(base), range(1, base+1)))
>>> d.reindex(index=[np.nan, 1])
NaN    NaN
1.0    0.0
dtype: float64


>>> # unexpected behavior
>>> base = 101
>>> d = pd.Series(range(base), index=pd.IntervalIndex.from_arrays(range(base), range(1, base+1)))
>>> d.reindex(index=[np.nan, 1])
NaN    49
1.0     0
dtype: int64


>>> # varying size of index
>>> values = []
>>> for base in range(1, 200):
   ...:      d = pd.Series(range(base), index=pd.IntervalIndex.from_arrays(range(base), range(1, base+1)))
   ...:      value = d.reindex(index=[np.nan])
   ...: 
   ...:      values.append(value.iloc[0])
>>> print(values)
[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 49, 50, 50, 51, 51, 52, 52, 53, 53, 54, 54, 55, 55, 56, 56, 57, 57, 58, 58, 59, 59, 60, 60, 61, 61, 62, 62, 63, 63, 64, 64, 65, 65, 66, 66, 67, 67, 68, 68, 69, 69, 70, 70, 71, 71, 72, 72, 73, 73, 74, 74, 75, 75, 76, 76, 77, 77, 78, 78, 79, 79, 80, 80, 81, 81, 82, 82, 83, 83, 84, 84, 85, 85, 86, 86, 87, 87, 88, 88, 89, 89, 90, 90, 91, 91, 92, 92, 93, 93, 94, 94, 95, 95, 96, 96, 97, 97, 98, 98]
```

### Issue Description

I am using interval indices and using `.reindex`  to index with missing labels (as suggested in the [docs](https://pandas.pydata.org/docs/user_guide/indexing.html?highlight=indexing%20list%20missing%20labels#indexing-with-list-with-missing-labels-is-deprecated)). However, it looks like if there are more than 100 rows in the index, when I `.reindex` with a missing key, i.e. `np.nan`, `.reindex` will return an unexpected non-null value. Since `np.nan` is not defined in the index, I would expect `.reindex` to return `np.nan`. But instead, it looks like value of a middle key is returned. When there are 100 rows or fewer, when I pass `np.nan` I get back what I expect, `np.nan`

### Expected Behavior

```python
>>> import pandas as pd
>>> import numpy as np

>>> # expected behavior, no change
>>> base = 100
>>> d = pd.Series(range(base), index=pd.IntervalIndex.from_arrays(range(base), range(1, base+1)))
>>> d.reindex(index=[np.nan, 1])
NaN    NaN
1.0    0.0
dtype: float64

>>> # unexpected behavior, changed to return `np.nan` when `np.nan` is passed in
>>> base = 101
>>> d = pd.Series(range(base), index=pd.IntervalIndex.from_arrays(range(base), range(1, base+1)))
>>> d.reindex(index=[np.nan, 1])
NaN    NaN
1.0    0.0
dtype: float64

>>> # varying size of index, changed to always return `np.nan` when `np.nan` is passed, regardless of the size of the index
>>> values = []
>>> for base in range(1, 200):
   ...:      d = pd.Series(range(base), index=pd.IntervalIndex.from_arrays(range(base), range(1, base+1)))
   ...:      value = d.reindex(index=[np.nan])
   ...: 
   ...:      values.append(value.iloc[0])
>>> print(values)
[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]

```

### Installed Versions

<details>

INSTALLED VERSIONS                                                                                                                    
------------------
commit           : 2e218d10984e9919f0296931d92ea851c6a6faf5
python           : 3.10.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-197-generic
Version          : #208-Ubuntu SMP Tue Nov 1 17:23:37 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.5.3
numpy            : 1.24.2
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 65.6.3
pip              : 23.0.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.11.0
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by raj-thapa**
- [x] closes #51826
- [x] [Tests added and passed]
- [x] All [code checks passed]

The test added in this issue is failing on Linux-32-bit. This is caused by another issue as discussed in #23440. The description of the the error is also described below in the comments. The test is marked xfail for 32-bit system in this PR.

### CODE REVIEW

File: pandas/tests/indexing/interval/test_interval.py [hunk 1]
```diff
@@ -172,3 +172,20 @@ def test_mi_intervalindex_slicing_with_scalar(self):
...
+            (101, Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)),
+            (1010, Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)),
+        ],
+    )
+    def test_reindex_behavior_with_interval_index(self, base, expected_result):
+        # GH 51826
+        left = np.arange(base)
+        right = np.arange(1, base + 1)
+        d = Series(range(base), index=IntervalIndex.from_arrays(left, right))
+        result = d.reindex(index=[np.nan, 1.0])
```
Comment (mroeschke) [inline pandas/tests/indexing/interval/test_interval.py:190]:
```suggestion
        ser = Series(range(base), index=IntervalIndex.from_arrays(left, right))
        result = ser.reindex(index=[np.nan, 1.0])
```

File: pandas/tests/indexing/interval/test_interval.py [hunk 2]
```diff
@@ -172,3 +172,20 @@ def test_mi_intervalindex_slicing_with_scalar(self):
         )
         expected = Series([1, 6, 2, 8, 7], index=expected_index, name="value")
         tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "base, expected_result",
+        [
+            (10, Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)),
```
Comment (mroeschke) [inline pandas/tests/indexing/interval/test_interval.py:179]: Looks like `expected_result` can be moved into the test body

**Review by mroeschke — COMMENTED** [review]
looks like the 32 bit tests are failing

**Comment by raj-thapa** [comment]
I have tried to enforce the 64-bit integers on the index arrays.
```
def test_interval_index_reindex_behavior(base, expected_result):
    # GH 51826
    left = np.arange(base, dtype=np.int64)
    right = np.arange(1, base + 1, dtype=np.int64)
    d = Series(
        range(base),
        index=IntervalIndex.from_arrays(left, right)
    )
    result = d.reindex(index=[np.nan, 1.0])
    tm.assert_series_equal(result, expected_result)
 ```
 However, the 32 bit tests still seems to fail. The specific error is 
 ```
 FAILED pandas/tests/indexing/interval/test_interval.py::TestIntervalIndexInsideMultiIndex::test_interval_index_reindex_behavior[101-expected_result2] - TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
FAILED pandas/tests/indexing/interval/test_interval.py::TestIntervalIndexInsideMultiIndex::test_interval_index_reindex_behavior[1010-expected_result3] - TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
```

And the error also points to `line:196` on `pandas/_libs/intervaltree.pxi` which is this return statement.
```
cdef take(ndarray source, ndarray indices):
    """Take the given positions from a 1D ndarray
    """
    return PyArray_Take(source, indices, 0)
```
 Do anyone have ideas on solving this unexpected casting issue?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @raj-thapa

**Pull request merged**